### PR TITLE
Stream audio

### PR DESCRIPTION
**Pull request by sigmaSd**
It works but there are different problems:

1-  it spins the cpu a lot, needs investigation, this is the main problem

2- it uses wavy crate, cpal crate would've been better for android and windows support but I can't even test cpal because it doesn't work on my pc, wavy has android and windows support planned though.

3- couple of minor stuff: It doesn't handle multiple streamers correctly, and this feature should be also gated

### CODE REVIEW

**Comment by lemunozm** [comment]
Regarding point 1. From message-io, when the CPU goes up it could be because of the active waiting in sending methods. `message-io` enters into this "active waiting" when the receiver has filled its input buffer and can not receive more data until process what it already has. Usually running this in release mode is enough for reading the packet faster than a sender can send in "localhost" (and for that, any scenario). I will investigate why the receiver fills its buffer because seems related to it.

**Comment by lemunozm** [comment]
Regarding point 1. The receiver is doing a lot of work here https://github.com/lemunozm/termchat/pull/39/files#diff-87619baf236177a7a9fc38150961151b7583b6f2f082ade75915f3ac4ca49c41R303 because it is always receiving data because the sender is sending without a rate. I think the problem comes from https://github.com/lemunozm/termchat/pull/39/files#diff-35d8224790c042925dbb9a36a948deddd5111e352bde21ec476b78d3876a8576R70 that most of the cases return 0 which implies wake up the executor on the receiver.

In other words: the sender never stops sending audio (100% CPU in the sender), and the receiver collapse with this amount of messages (100% CPU in the receiver). 

I think that being able to offer some kind of delay at Processing enum, like: `Processing::Partial(Duration::from_millis(1))` in order to send the audio in grater chunks, would reduce the CPU at the sender and avoid saturate the receiver. I will add a PR with this architectural feature.

**Comment by sigmaSd** [comment]
Pr updated.

Now its seems that the performance is acceptable(especially in release mode).

File: src/application.rs [hunk 1]
```diff
@@ -200,6 +205,10 @@ impl<'a> Application<'a> {
                     self.state.windows.remove(&endpoint);
                 }
             }
+            NetMessage::StreamAudio(audio) => match audio {
+                Some(audio) => self.state.pulse_audio(audio),
```
Comment (sigmaSd) [inline src/application.rs]: I'm curious if this handles already multi user stream, or do i need to tag each users stream (live video stream).

@lemunozm Are you able to test this? currently I don't a second linux pc.

Basically the question is what happens when different audio sources are merged, does the audio remain understandable?

Comment (lemunozm) [reply src/application.rs]: I don't have another Linux either 😢.
Seeing how it works I suppose that the audio is mixed, but I do not see it as a problem.
The problem could be with sound in two directions. I think it will be coupled.

Maybe it could be cool that it behaves more like a WhatsApp audio in the sense of if I send an audio, the receiver could choose when to listen to it. Even it can listen to the same audio several times. This feature would imply navigating along the conversation allowing to select an audio to listen to.

Comment (sigmaSd) [reply src/application.rs]: The whatsup idea is cool, it could be the same for sending pictures and files.

It could be for now simply different commands, stream-audio/stream-video | send-audio/send-video, Currently we're finishing the first set, the second set should be possible after some ui work.

**Comment by sigmaSd** [comment]
I think this should be closed now, hopefully someone does a better implementation

**Comment by lemunozm** [comment]
Hi @sigmaSd, I had no opportunity to test it. But if you consider the feature is working, I'm ok with adding it.
